// @polkadot/api/Base.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _eventemitter = _interopRequireDefault(require("eventemitter3"));

var _rxjs = require("rxjs");

var _operators = require("rxjs/operators");

var _apiDerive = _interopRequireDefault(require("@polkadot/api-derive"));

var _fromMetadata = _interopRequireDefault(require("@polkadot/api-metadata/consts/fromMetadata"));

var _fromMetadata2 = _interopRequireDefault(require("@polkadot/api-metadata/extrinsics/fromMetadata"));

var _fromMetadata3 = _interopRequireDefault(require("@polkadot/api-metadata/storage/fromMetadata"));

var _rpcCore = _interopRequireDefault(require("@polkadot/rpc-core"));

var _rpcProvider = require("@polkadot/rpc-provider");

var _types = require("@polkadot/types");

var _Linkage = require("@polkadot/types/codec/Linkage");

var _constants = require("@polkadot/types/primitive/Extrinsic/constants");

var srmlTypes = _interopRequireWildcard(require("@polkadot/types/srml/definitions"));

var _util = require("@polkadot/util");

var _utilCrypto = require("@polkadot/util-crypto");

var _SubmittableExtrinsic = _interopRequireDefault(require("./SubmittableExtrinsic"));

var _decorate = require("./util/decorate");

// Copyright 2017-2019 @polkadot/api authors & contributors
// This software may be modified and distributed under the terms
// of the Apache-2.0 license. See the LICENSE file for details.
const INIT_ERROR = "Api needs to be initialised before using, listen on 'ready'";
const KEEPALIVE_INTERVAL = 15000;
const l = (0, _util.logger)('api/decorator');
let pkgJson;

try {
  pkgJson = require('./package.json');
} catch (error) {
  // development environment
  pkgJson = {
    name: '@polkadot/api',
    version: '-'
  };
}
/**
 * Put the `this.onCall` function of ApiRx here, because it is needed by
 * `api._rx`.
 */


function rxDecorateMethod(method) {
  return method;
}

class ApiBase {
  /**
   * @description Create an instance of the class
   *
   * @param options Options object to create API instance or a Provider instance
   *
   * @example
   * <BR>
   *
   * ```javascript
   * import Api from '@polkadot/api/promise';
   *
   * const api = new Api().isReady();
   *
   * api.rpc.subscribeNewHead((header) => {
   *   console.log(`new block #${header.blockNumber.toNumber()}`);
   * });
   * ```
   */
  constructor() {
    let provider = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    let type = arguments.length > 1 ? arguments[1] : undefined;
    this._consts = void 0;
    this._derive = void 0;
    this._eventemitter = void 0;
    this._extrinsics = void 0;
    this._extrinsicType = _constants.DEFAULT_VERSION;
    this._genesisHash = void 0;
    this._isConnected = void 0;
    this._isReady = false;
    this._options = void 0;
    this._query = void 0;
    this._queryMulti = void 0;
    this._rpc = void 0;
    this._rpcCore = void 0;
    this._runtimeMetadata = void 0;
    this._runtimeVersion = void 0;
    this._rx = {};
    this._type = void 0;
    const options = (0, _util.isObject)(provider) && (0, _util.isFunction)(provider.send) ? {
      provider
    } : provider;
    const thisProvider = options.source ? options.source._rpcCore.provider.clone() : options.provider || new _rpcProvider.WsProvider();
    this._options = options;
    this._type = type;
    this._eventemitter = new _eventemitter.default();
    this._rpcCore = new _rpcCore.default(thisProvider);
    this._isConnected = new _rxjs.BehaviorSubject(this._rpcCore.provider.isConnected());
    (0, _util.assert)(this.hasSubscriptions, 'Api can only be used with a provider supporting subscriptions');
    this._rpc = this.decorateRpc(this._rpcCore, this.decorateMethod);
    this._rx.rpc = this.decorateRpc(this._rpcCore, rxDecorateMethod);
    this._queryMulti = this.decorateMulti(this.decorateMethod);
    this._rx.queryMulti = this.decorateMulti(rxDecorateMethod);
    this._rx.signer = options.signer; // we only re-register the types (global) if this is not a cloned instance

    if (!options.source) {
      // first register the definitions we have, i.e. those where there are no type classes
      Object.values(srmlTypes).forEach((_ref) => {
        let {
          types
        } = _ref;
        return this.registerTypes(types);
      }); // next register all the user types

      this.registerTypes(options.types);
    }

    this.init();
  }
  /**
   * @description  Returns th version of extrinsics in-use on this chain
   */


  get extrinsicVersion() {
    return this._extrinsicType;
  }
  /**
   * @description Contains the genesis Hash of the attached chain. Apart from being useful to determine the actual chain, it can also be used to sign immortal transactions.
   */


  get genesisHash() {
    (0, _util.assert)(!(0, _util.isUndefined)(this._genesisHash), INIT_ERROR);
    return this._genesisHash;
  }
  /**
   * @description `true` when subscriptions are supported
   */


  get hasSubscriptions() {
    return this._rpcCore.provider.hasSubscriptions;
  }
  /**
   * @description The library information name & version (from package.json)
   */


  get libraryInfo() {
    return "".concat(pkgJson.name, " v").concat(pkgJson.version);
  }
  /**
   * @description Yields the current attached runtime metadata. Generally this is only used to construct extrinsics & storage, but is useful for current runtime inspection.
   */


  get runtimeMetadata() {
    (0, _util.assert)(!(0, _util.isUndefined)(this._runtimeMetadata), INIT_ERROR);
    return this._runtimeMetadata;
  }
  /**
   * @description Contains the version information for the current runtime.
   */


  get runtimeVersion() {
    (0, _util.assert)(!(0, _util.isUndefined)(this._runtimeVersion), INIT_ERROR);
    return this._runtimeVersion;
  }
  /**
   * @description The type of this API instance, either 'rxjs' or 'promise'
   */


  get type() {
    return this._type;
  }
  /**
   * @description Set an external signer which will be used to sign extrinsic when account passed in is not KeyringPair
   */


  setSigner(signer) {
    this._rx.signer = signer;
  }
  /**
   * @description Derived results that are injected into the API, allowing for combinations of various query results.
   *
   * @example
   * <BR>
   *
   * ```javascript
   * api.derive.chain.bestNumber((number) => {
   *   console.log('best number', number);
   * });
   * ```
   */


  get derive() {
    (0, _util.assert)(!(0, _util.isUndefined)(this._derive), INIT_ERROR);
    return this._derive;
  }
  /**
   * @description Contains the parameter types (constants) of all modules.
   *
   * The values are instances of the appropriate type and are accessible using `section`.`constantName`,
   *
   * @example
   * <BR>
   *
   * ```javascript
   * console.log(api.consts.democracy.enactmentPeriod.toString())
   * ```
   */


  get consts() {
    (0, _util.assert)(!(0, _util.isUndefined)(this._consts), INIT_ERROR);
    return this._consts;
  }
  /**
   * @description Contains all the chain state modules and their subsequent methods in the API. These are attached dynamically from the runtime metadata.
   *
   * All calls inside the namespace, is denoted by `section`.`method` and may take an optional query parameter. As an example, `api.query.timestamp.now()` (current block timestamp) does not take parameters, while `api.query.system.accountNonce(<accountId>)` (retrieving the associated nonce for an account), takes the `AccountId` as a parameter.
   *
   * @example
   * <BR>
   *
   * ```javascript
   * api.query.balances.freeBalance(<accountId>, (balance) => {
   *   console.log('new balance', balance);
   * });
   * ```
   */


  get query() {
    (0, _util.assert)(!(0, _util.isUndefined)(this._query), INIT_ERROR);
    return this._query;
  }
  /**
   * @description Allows for the querying of multiple storage entries and the combination thereof into a single result. This is a very optimal way to make multiple queries since it only makes a single connection to the node and retrieves the data over one subscription.
   *
   * @example
   * <BR>
   *
   * ```javascript
   * const unsub = await api.queryMulti(
   *   [
   *     // you can include the storage without any parameters
   *     api.query.balances.totalIssuance,
   *     // or you can pass parameters to the storage query
   *     [api.query.balances.freeBalance, '5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY']
   *   ],
   *   ([existential, balance]) => {
   *     console.log(`You have ${balance.sub(existential)} more than the existential deposit`);
   *
   *     unsub();
   *   }
   * );
   * ```
   */


  get queryMulti() {
    return this._queryMulti;
  }
  /**
   * @description Contains all the raw rpc sections and their subsequent methods in the API as defined by the jsonrpc interface definitions. Unlike the dynamic `api.query` and `api.tx` sections, these methods are fixed (although extensible with node upgrades) and not determined by the runtime.
   *
   * RPC endpoints available here allow for the query of chain, node and system information, in addition to providing interfaces for the raw queries of state (usine known keys) and the submission of transactions.
   *
   * @example
   * <BR>
   *
   * ```javascript
   * api.rpc.chain.subscribeNewHead((header) => {
   *   console.log('new header', header);
   * });
   * ```
   */


  get rpc() {
    return this._rpc;
  }
  /**
   * @description Contains all the extrinsic modules and their subsequent methods in the API. It allows for the construction of transactions and the submission thereof. These are attached dynamically from the runtime metadata.
   *
   * @example
   * <BR>
   *
   * ```javascript
   * api.tx.balances
   *   .transfer(<recipientId>, <balance>)
   *   .signAndSend(<keyPair>, ({status}) => {
   *     console.log('tx status', status.asFinalized.toHex());
   *   });
   * ```
   */


  get tx() {
    (0, _util.assert)(!(0, _util.isUndefined)(this._extrinsics), INIT_ERROR);
    return this._extrinsics;
  }
  /**
   * @description Disconnect from the underlying provider, halting all comms
   */


  disconnect() {
    this._rpcCore.disconnect();
  }
  /**
   * @description Attach an eventemitter handler to listen to a specific event
   *
   * @param type The type of event to listen to. Available events are `connected`, `disconnected`, `ready` and `error`
   * @param handler The callback to be called when the event fires. Depending on the event type, it could fire with additional arguments.
   *
   * @example
   * <BR>
   *
   * ```javascript
   * api.on('connected', (): void => {
   *   console.log('API has been connected to the endpoint');
   * });
   *
   * api.on('disconnected', (): void => {
   *   console.log('API has been disconnected from the endpoint');
   * });
   * ```
   */


  on(type, handler) {
    this._eventemitter.on(type, handler);

    return this;
  }
  /**
   * @description Remove the given eventemitter handler
   *
   * @param type The type of event the callback was attached to. Available events are `connected`, `disconnected`, `ready` and `error`
   * @param handler The callback to unregister.
   *
   * @example
   * <BR>
   *
   * ```javascript
   * const handler = (): void => {
   *  console.log('Connected !);
   * };
   *
   * // Start listening
   * api.on('connected', handler);
   *
   * // Stop listening
   * api.off('connected', handler);
   * ```
   */


  off(type, handler) {
    this._eventemitter.removeListener(type, handler);

    return this;
  }
  /**
   * @description Attach an one-time eventemitter handler to listen to a specific event
   *
   * @param type The type of event to listen to. Available events are `connected`, `disconnected`, `ready` and `error`
   * @param handler The callback to be called when the event fires. Depending on the event type, it could fire with additional arguments.
   *
   * @example
   * <BR>
   *
   * ```javascript
   * api.once('connected', (): void => {
   *   console.log('API has been connected to the endpoint');
   * });
   *
   * api.once('disconnected', (): void => {
   *   console.log('API has been disconnected from the endpoint');
   * });
   * ```
   */


  once(type, handler) {
    this._eventemitter.once(type, handler);

    return this;
  }
  /**
   * @description Register additional user-defined of chain-specific types in the type registry
   */


  registerTypes(types) {
    if (types) {
      (0, _types.getTypeRegistry)().register(types);
    }
  }
  /**
   * This is the one and only method concrete children classes need to implement.
   * It's a higher-order function, which takes one argument
   * `method: Method extends (...args: any[]) => Observable<any>`
   * (and one optional `options`), and should return the user facing method.
   * For example:
   * - For ApiRx, `decorateMethod` should just be identity, because the input
   * function is already an Observable
   * - For ApiPromise, `decorateMethod` should return a function that takes all
   * the parameters from `method`, adds an optional `callback` argument, and
   * returns a Promise.
   *
   * We could easily imagine other user-facing interfaces, which are simply
   * implemented by transforming the Observable to Stream/Iterator/Kefir/Bacon
   * via `deocrateMethod`.
   */


  emit(type) {
    for (var _len = arguments.length, args = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
      args[_key - 1] = arguments[_key];
    }

    this._eventemitter.emit(type, ...args);
  }

  init() {
    let healthTimer = null;

    this._rpcCore.provider.on('disconnected', () => {
      this.emit('disconnected');

      this._isConnected.next(false);

      if (healthTimer) {
        clearInterval(healthTimer);
        healthTimer = null;
      }
    });

    this._rpcCore.provider.on('error', error => {
      this.emit('error', error);
    });

    this._rpcCore.provider.on('connected', async () => {
      this.emit('connected');

      this._isConnected.next(true);

      try {
        const [hasMeta, cryptoReady] = await Promise.all([this.loadMeta(), (0, _utilCrypto.cryptoWaitReady)()]);

        if (hasMeta && !this._isReady && cryptoReady) {
          this._isReady = true;
          this.emit('ready', this);
        }

        healthTimer = setInterval(() => {
          this._rpcCore.system.health().toPromise().catch(() => {// ignore
          });
        }, KEEPALIVE_INTERVAL);
      } catch (_error) {
        const error = new Error("FATAL: Unable to initialize the API: ".concat(_error.message));
        l.error(error);
        this.emit('error', error);
      }
    });
  }

  async loadMeta() {
    const {
      metadata = {}
    } = this._options; // only load from on-chain if we are not a clone (default path), alternatively
    // just use the values from the source instance provided

    if (!this._options.source || !this._options.source._isReady) {
      [this._genesisHash, this._runtimeVersion] = await Promise.all([this._rpcCore.chain.getBlockHash(0).toPromise(), this._rpcCore.chain.getRuntimeVersion().toPromise()]);
      const metadataKey = "".concat(this._genesisHash, "-").concat(this._runtimeVersion.specVersion);

      if (metadataKey in metadata) {
        this._runtimeMetadata = new _types.Metadata(metadata[metadataKey]);
      } else {
        this._runtimeMetadata = await this._rpcCore.state.getMetadata().toPromise();
      } // get unique types & validate


      this.runtimeMetadata.getUniqTypes(false);
    } else {
      this._extrinsicType = this._options.source.extrinsicVersion;
      this._runtimeMetadata = this._options.source.runtimeMetadata;
      this._runtimeVersion = this._options.source.runtimeVersion;
      this._genesisHash = this._options.source.genesisHash;
    }

    const extrinsics = (0, _fromMetadata2.default)(this.runtimeMetadata);
    const storage = (0, _fromMetadata3.default)(this.runtimeMetadata);
    const constants = (0, _fromMetadata.default)(this.runtimeMetadata); // only inject if we are not a clone (global init)

    if (!this._options.source) {
      _types.Event.injectMetadata(this.runtimeMetadata);

      _types.Method.injectMethods(extrinsics); // detect the extrinsic version in-use based on the last block


      const lastBlock = await this._rpcCore.chain.getBlock().toPromise();
      this._extrinsicType = lastBlock.block.extrinsics[0].type;
    }

    this._extrinsics = this.decorateExtrinsics(extrinsics, this.decorateMethod);
    this._query = this.decorateStorage(storage, this.decorateMethod);
    this._consts = constants;
    this._rx.extrinsicType = this._extrinsicType;
    this._rx.genesisHash = this._genesisHash;
    this._rx.runtimeVersion = this._runtimeVersion;
    this._rx.tx = this.decorateExtrinsics(extrinsics, rxDecorateMethod);
    this._rx.query = this.decorateStorage(storage, rxDecorateMethod);
    this._rx.consts = constants;
    this._derive = this.decorateDerive(this._rx, this.decorateMethod);
    return true;
  }

  decorateFunctionMeta(input, output) {
    output.meta = input.meta;
    output.method = input.method;
    output.section = input.section;
    output.toJSON = input.toJSON;

    if (input.callIndex) {
      output.callIndex = input.callIndex;
    }

    return output;
  }

  decorateRpc(rpc, decorateMethod) {
    return ['author', 'chain', 'state', 'system'].reduce((result, _sectionName) => {
      const sectionName = _sectionName;
      result[sectionName] = Object.keys(rpc[sectionName]).reduce((section, methodName) => {
        const method = rpc[sectionName][methodName];
        section[methodName] = decorateMethod(method, {
          methodName
        });
        return section;
      }, {});
      return result;
    }, {});
  }

  decorateMulti(decorateMethod) {
    return decorateMethod(calls => {
      const mapped = calls.map(arg => // the input is a QueryableStorageEntry, convert to StorageEntry
      Array.isArray(arg) ? [arg[0].creator, ...arg.slice(1)] : [arg.creator]);
      return this._rpcCore.state.subscribeStorage(mapped);
    });
  }

  decorateExtrinsics(extrinsics, decorateMethod) {
    const creator = value => (0, _SubmittableExtrinsic.default)(this.type, this._rx, decorateMethod, value);

    return Object.keys(extrinsics).reduce((result, sectionName) => {
      const section = extrinsics[sectionName];
      result[sectionName] = Object.keys(section).reduce((result, methodName) => {
        result[methodName] = this.decorateExtrinsicEntry(section[methodName], decorateMethod);
        return result;
      }, {});
      return result;
    }, creator);
  }

  decorateExtrinsicEntry(method, decorateMethod) {
    var _this = this;

    const decorated = function decorated() {
      return (0, _SubmittableExtrinsic.default)(_this.type, _this._rx, decorateMethod, method(...arguments));
    };

    return this.decorateFunctionMeta(method, decorated);
  }

  decorateStorage(storage, decorateMethod) {
    return Object.keys(storage).reduce((result, sectionName) => {
      const section = storage[sectionName];
      result[sectionName] = Object.keys(section).reduce((result, methodName) => {
        result[methodName] = this.decorateStorageEntry(section[methodName], decorateMethod);
        return result;
      }, {});
      return result;
    }, {});
  }

  decorateStorageEntry(creator, decorateMethod) {
    var _this2 = this;

    const decorated = creator.headKey ? this.decorateStorageEntryLinked(creator, decorateMethod) : decorateMethod(function () {
      for (var _len2 = arguments.length, args = new Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
        args[_key2] = arguments[_key2];
      }

      return _this2._rpcCore.state // Unfortunately for one-shot calls we also use .subscribeStorage here
      .subscribeStorage([creator.meta.type.isDoubleMap ? [creator, args] : [creator, ...args]]).pipe( // state_storage returns an array of values, since we have just subscribed to
      // a single entry, we pull that from the array and return it as-is
      (0, _operators.map)(result => result[0]));
    }, {
      methodName: creator.method
    });
    decorated.creator = creator;
    decorated.at = decorateMethod((hash, arg1, arg2) => this._rpcCore.state.getStorage(creator.meta.type.isDoubleMap ? [creator, [arg1, arg2]] : [creator, arg1], hash));
    decorated.hash = decorateMethod((arg1, arg2) => this._rpcCore.state.getStorageHash(creator.meta.type.isDoubleMap ? [creator, [arg1, arg2]] : [creator, arg1]));

    decorated.key = (arg1, arg2) => (0, _util.u8aToHex)((0, _util.compactStripLength)(creator(creator.meta.type.isDoubleMap ? [arg1, arg2] : arg1))[1]); // When using double map storage function, user need to path double map key as an array


    decorated.multi = decorateMethod(args => this._rpcCore.state.subscribeStorage(args.map(arg => [creator, arg])));
    decorated.size = decorateMethod((arg1, arg2) => this._rpcCore.state.getStorageSize(creator.meta.type.isDoubleMap ? [creator, [arg1, arg2]] : [creator, arg1]));
    return this.decorateFunctionMeta(creator, decorated);
  }

  decorateStorageEntryLinked(method, decorateMethod) {
    const result = new Map();
    let subject;
    let head = null; // retrieve a value based on the key, iterating if it has a next entry. Since
    // entries can be re-linked in the middle of a list, we subscribe here to make
    // sure we catch any updates, no matter the list position

    const getNext = key => {
      return this._rpcCore.state.subscribeStorage([[method, key]]).pipe((0, _operators.switchMap)((_ref2) => {
        let [data] = _ref2;
        const linkage = data[1];
        result.set(key, data); // iterate from this key to the children, constructing
        // entries for all those found and available

        if (linkage.next.isSome) {
          return getNext(linkage.next.unwrap());
        }

        const keys = [];
        const values = [];
        let nextKey = head; // loop through the results collected, starting at the head an re-creating
        // the list. Our map may have old entries, based on the linking these will
        // not be returned in the final result

        while (nextKey) {
          const entry = result.get(nextKey);

          if (!entry) {
            break;
          }

          const [item, linkage] = entry;
          keys.push(nextKey);
          values.push(item);
          nextKey = linkage.next && linkage.next.unwrapOr(null);
        }

        const nextResult = values.length ? new _Linkage.LinkageResult([keys[0].constructor, keys], [values[0].constructor, values]) : new _Linkage.LinkageResult([_types.Null, []], [_types.Null, []]); // we set our result into a subject so we have a single observable for
        // which the value changes over time. Initially create, follow-up next

        if (subject) {
          subject.next(nextResult);
        } else {
          subject = new _rxjs.BehaviorSubject(nextResult);
        }

        return subject;
      }));
    }; // this handles the case where the head changes effectively, i.e. a new entry
    // appears at the top of the list, the new getNext gets kicked off


    return decorateMethod(() => this._rpcCore.state.subscribeStorage([method.headKey]).pipe((0, _operators.switchMap)((_ref3) => {
      let [key] = _ref3;
      head = key;
      return getNext(key);
    })));
  } // FIXME I have no idea how to get this done
  // eslint-disable-next-line @typescript-eslint/explicit-function-return-type


  decorateDerive(apiRx, decorateMethod) {
    // Pull in derive from api-derive
    const derive = (0, _apiDerive.default)(apiRx, this._options.derives);
    return (0, _decorate.decorateSections)(derive, decorateMethod);
  }

}

exports.default = ApiBase;
