// @polkadot/ui-settings/defaults.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UITHEMES = exports.UITHEME_DEFAULT = exports.UIMODES = exports.UIMODE_DEFAULT = exports.PREFIXES = exports.PREFIX_DEFAULT = exports.LOCKING = exports.LOCKING_DEFAULT = exports.LANGUAGES = exports.LANGUAGE_DEFAULT = exports.ENDPOINTS = exports.ENDPOINT_DEFAULT = exports.CRYPTOS = void 0;
// Copyright 2017-2019 @polkadot/ui-settings authors & contributors
// This software may be modified and distributed under the terms
// of the Apache-2.0 license. See the LICENSE file for details.
// matches https://polkadot.js.org & https://*.polkadot.io
// tslint:disable-next-line
const isPolkadot = typeof window !== 'undefined' && window.location.host.indexOf('polkadot') !== -1;
const WSS_NODES = {
  parity: {
    hosted: 'hosted by Parity',
    nodes: {
      alex: 'wss://poc3-rpc.polkadot.io/',
      fir: 'wss://substrate-rpc.parity.io/'
    }
  },
  unfra: {
    hosted: 'hosted by Centrality UNfrastructure',
    nodes: {
      alex: 'wss://alex.unfrastructure.io/public/ws'
    }
  }
};
const LANGUAGE_DEFAULT = 'default';
exports.LANGUAGE_DEFAULT = LANGUAGE_DEFAULT;
const LOCKING_DEFAULT = 'session';
exports.LOCKING_DEFAULT = LOCKING_DEFAULT;
const CRYPTOS = [{
  text: 'Edwards (ed25519)',
  value: 'ed25519'
}, {
  text: 'Schnorrkel (sr25519)',
  value: 'sr25519'
}];
exports.CRYPTOS = CRYPTOS;
const ENDPOINTS = [{
  text: "Alexander (Polkadot, ".concat(WSS_NODES.parity.hosted, ")"),
  value: WSS_NODES.parity.nodes.alex
}, {
  text: "Alexander (Polkadot, ".concat(WSS_NODES.unfra.hosted, ")"),
  value: WSS_NODES.unfra.nodes.alex
}, {
  text: "Flaming Fir (Substrate, ".concat(WSS_NODES.parity.hosted, ")"),
  value: WSS_NODES.parity.nodes.fir
}, {
  text: 'Local Node (127.0.0.1:9944)',
  value: 'ws://127.0.0.1:9944/'
}];
exports.ENDPOINTS = ENDPOINTS;
const LANGUAGES = [{
  text: 'Default browser language (auto-detect)',
  value: LANGUAGE_DEFAULT
}];
exports.LANGUAGES = LANGUAGES;
const LOCKING = [{
  text: 'Once per session',
  value: 'session'
}, {
  text: 'On each transaction',
  value: 'tx'
}];
exports.LOCKING = LOCKING;
const PREFIXES = [{
  text: 'Default for the connected node',
  value: -1
}, {
  text: 'Substrate (development)',
  value: 42
}, {
  text: 'Kusama (canary)',
  value: 2
}, {
  text: 'Polkadot (live)',
  value: 0
}];
exports.PREFIXES = PREFIXES;
const UIMODES = [{
  value: 'full',
  text: 'Fully featured'
}, {
  value: 'light',
  text: 'Basic features only'
}];
exports.UIMODES = UIMODES;
const UITHEMES = [{
  value: 'polkadot',
  text: 'Polkadot'
}, {
  value: 'substrate',
  text: 'Substrate'
}];
exports.UITHEMES = UITHEMES;
const ENDPOINT_DEFAULT = isPolkadot ? WSS_NODES.parity.nodes.alex : WSS_NODES.parity.nodes.fir;
exports.ENDPOINT_DEFAULT = ENDPOINT_DEFAULT;
const PREFIX_DEFAULT = -1;
exports.PREFIX_DEFAULT = PREFIX_DEFAULT;
const UITHEME_DEFAULT = isPolkadot ? 'polkadot' : 'substrate'; // tslint:disable-next-line

exports.UITHEME_DEFAULT = UITHEME_DEFAULT;
const UIMODE_DEFAULT = !isPolkadot && typeof window !== 'undefined' && window.location.host.indexOf('ui-light') !== -1 ? 'light' : 'full';
exports.UIMODE_DEFAULT = UIMODE_DEFAULT;
