// @polkadot/types/primitive/Extrinsic/v1/SignaturePayload.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _Struct = _interopRequireDefault(require("../../../codec/Struct"));

var _U8a = _interopRequireDefault(require("../../../codec/U8a"));

var _Hash = _interopRequireDefault(require("../../Hash"));

var _NonceCompact = _interopRequireDefault(require("../../../type/NonceCompact"));

var _ExtrinsicEra = _interopRequireDefault(require("../ExtrinsicEra"));

var _util = require("../util");

// Copyright 2017-2019 @polkadot/types authors & contributors
// This software may be modified and distributed under the terms
// of the Apache-2.0 license. See the LICENSE file for details.

/**
 * @name SignaturePayloadV1
 * @description
 * A signing payload for an [[Extrinsic]]. For the final encoding, it is variable length based
 * on the contents included
 *
 *   1-8 bytes: The Transaction Compact<Index/Nonce> as provided in the transaction itself.
 *   2+ bytes: The Function Descriptor as provided in the transaction itself.
 *   1/2 bytes: The Transaction Era as provided in the transaction itself.
 *   32 bytes: The hash of the authoring block implied by the Transaction Era and the current block.
 */
class SignaturePayloadV1 extends _Struct.default {
  constructor(value) {
    super({
      nonce: _NonceCompact.default,
      method: _U8a.default,
      era: _ExtrinsicEra.default,
      blockHash: _Hash.default
    }, value);
  }
  /**
   * @description The block [[Hash]] the signature applies to (mortal/immortal)
   */


  get blockHash() {
    return this.get('blockHash');
  }
  /**
   * @description The [[U8a]] contained in the payload
   */


  get method() {
    return this.get('method');
  }
  /**
   * @description The [[ExtrinsicEra]]
   */


  get era() {
    return this.get('era');
  }
  /**
   * @description The [[NonceCompact]]
   */


  get nonce() {
    return this.get('nonce');
  }
  /**
   * @description Sign the payload with the keypair
   */
  // eslint-disable-next-line @typescript-eslint/no-unused-vars


  sign(signerPair) {
    return (0, _util.sign)(signerPair, this.toU8a());
  }

}

exports.default = SignaturePayloadV1;
