// @polkadot/types/primitive/Extrinsic/SignaturePayload.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _util = require("@polkadot/util");

var _Base = _interopRequireDefault(require("../../codec/Base"));

var _BalanceCompact = _interopRequireDefault(require("../BalanceCompact"));

var _SignaturePayload = _interopRequireDefault(require("./v1/SignaturePayload"));

var _SignaturePayload2 = _interopRequireDefault(require("./v2/SignaturePayload"));

var _constants = require("./constants");

// Copyright 2017-2019 @polkadot/types authors & contributors
// This software may be modified and distributed under the terms
// of the Apache-2.0 license. See the LICENSE file for details.

/**
 * @name SignaturePayload
 * @description
 * A signing payload for an [[Extrinsic]]. For the final encoding, it is variable length based
 * on the contents included
 */
class SignaturePayload extends _Base.default {
  constructor(value) {
    let {
      version
    } = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
    super(SignaturePayload.decodeSignaturePayload(value, version));
  }

  static decodeSignaturePayload(value) {
    let version = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : _constants.DEFAULT_VERSION;

    if (value instanceof SignaturePayload) {
      return value.raw;
    }

    switch (version) {
      case 1:
        return new _SignaturePayload.default(value);

      case 2:
        return new _SignaturePayload2.default(value);

      default:
        throw new Error("Unsupported extrinsic version ".concat(version));
    }
  }
  /**
   * @description The block [[Hash]] the signature applies to (mortal/immortal)
   */


  get blockHash() {
    return this.raw.blockHash;
  }
  /**
   * @description The [[U8a]] contained in the payload
   */


  get method() {
    return this.raw.method;
  }
  /**
   * @description The [[ExtrinsicEra]]
   */


  get era() {
    return this.raw.era;
  }
  /**
   * @description The [[NonceCompact]]
   */


  get nonce() {
    return this.raw.nonce;
  }
  /**
   * @description The [[BalanceCompact]]
   */


  get tip() {
    return this.raw.tip || new _BalanceCompact.default(0);
  }
  /**
   * @description Compares the value of the input to see if there is a match
   */


  eq(other) {
    return this.raw.eq(other);
  }
  /**
   * @description Sign the payload with the keypair
   */


  sign(signerPair) {
    const signature = this.raw.sign(signerPair); // This is extensible, so we could quite readily extend to send back extra
    // information, such as for instance the payload, i.e. `payload: this.toHex()`
    // For the case here we sign via the extrinsic, we ignore the return, so generally
    // thisis applicable for external signing

    return {
      signature: (0, _util.u8aToHex)(signature)
    };
  }
  /**
   * @description Converts the Object to JSON, typically used for RPC transfers
   */


  toJSON() {
    return this.toHex();
  }
  /**
   * @description Returns the string representation of the value
   */


  toString() {
    return this.toHex();
  }

}

exports.default = SignaturePayload;
