// @polkadot/types/srml/definitions.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "session", {
  enumerable: true,
  get: function get() {
    return _definitions.default;
  }
});
Object.defineProperty(exports, "grandpa", {
  enumerable: true,
  get: function get() {
    return _definitions2.default;
  }
});
Object.defineProperty(exports, "babe", {
  enumerable: true,
  get: function get() {
    return _definitions3.default;
  }
});
Object.defineProperty(exports, "contracts", {
  enumerable: true,
  get: function get() {
    return _definitions4.default;
  }
});
Object.defineProperty(exports, "elections", {
  enumerable: true,
  get: function get() {
    return _definitions5.default;
  }
});
Object.defineProperty(exports, "imOnline", {
  enumerable: true,
  get: function get() {
    return _definitions6.default;
  }
});
Object.defineProperty(exports, "parachains", {
  enumerable: true,
  get: function get() {
    return _definitions7.default;
  }
});

var _definitions = _interopRequireDefault(require("./session/definitions"));

var _definitions2 = _interopRequireDefault(require("./grandpa/definitions"));

var _definitions3 = _interopRequireDefault(require("./babe/definitions"));

var _definitions4 = _interopRequireDefault(require("./contracts/definitions"));

var _definitions5 = _interopRequireDefault(require("./elections/definitions"));

var _definitions6 = _interopRequireDefault(require("./imOnline/definitions"));

var _definitions7 = _interopRequireDefault(require("./parachains/definitions"));
