// @polkadot/types/Metadata/v6/Storage.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "DoubleMapType", {
  enumerable: true,
  get: function get() {
    return _Storage.DoubleMapType;
  }
});
Object.defineProperty(exports, "MapType", {
  enumerable: true,
  get: function get() {
    return _Storage.MapType;
  }
});
Object.defineProperty(exports, "PlainType", {
  enumerable: true,
  get: function get() {
    return _Storage.PlainType;
  }
});
Object.defineProperty(exports, "StorageEntryMetadata", {
  enumerable: true,
  get: function get() {
    return _Storage.StorageFunctionMetadata;
  }
});
Object.defineProperty(exports, "StorageEntryMetadataValue", {
  enumerable: true,
  get: function get() {
    return _Storage.StorageFunctionMetadataValue;
  }
});
Object.defineProperty(exports, "StorageEntryModifier", {
  enumerable: true,
  get: function get() {
    return _Storage.StorageFunctionModifier;
  }
});
Object.defineProperty(exports, "StorageEntryType", {
  enumerable: true,
  get: function get() {
    return _Storage.StorageFunctionType;
  }
});

var _Storage = require("../v5/Storage");
