// @polkadot/types/primitive/Extrinsic/v2/ExtrinsicSignature.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _Struct = _interopRequireDefault(require("../../../codec/Struct"));

var _Address = _interopRequireDefault(require("../../Address"));

var _Signature = _interopRequireDefault(require("../../Signature"));

var _SignaturePayload = _interopRequireDefault(require("./SignaturePayload"));

var _constants = require("../constants");

var _ExtrinsicExtra = _interopRequireDefault(require("./ExtrinsicExtra"));

// Copyright 2017-2019 @polkadot/types authors & contributors
// This software may be modified and distributed under the terms
// of the Apache-2.0 license. See the LICENSE file for details.

/**
 * @name ExtrinsicSignature
 * @description
 * A container for the [[Signature]] associated with a specific [[Extrinsic]]
 */
class ExtrinsicSignatureV2 extends _Struct.default {
  constructor(value) {
    let {
      isSigned
    } = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
    super({
      signer: _Address.default,
      signature: _Signature.default,
      extra: _ExtrinsicExtra.default
    }, ExtrinsicSignatureV2.decodeExtrinsicSignature(value, isSigned));
  }

  static decodeExtrinsicSignature(value) {
    let isSigned = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;

    if (!value) {
      return _constants.EMPTY_U8A;
    } else if (value instanceof ExtrinsicSignatureV2) {
      return value;
    }

    return isSigned ? value : _constants.EMPTY_U8A;
  }
  /**
   * @description `true` if the signature is valid
   */


  get isSigned() {
    return !this.signature.isEmpty;
  }
  /**
   * @description Returns the extra extrinsic info
   */


  get extra() {
    return this.get('extra');
  }
  /**
   * @description The [[ExtrinsicEra]] (mortal or immortal) this signature applies to
   */


  get era() {
    return this.extra.era;
  }
  /**
   * @description The [[NonceCompact]] for the signature
   */


  get nonce() {
    return this.extra.nonce;
  }
  /**
   * @description The actuall [[Signature]] hash
   */


  get signature() {
    return this.get('signature');
  }
  /**
   * @description The [[Address]] that signed
   */


  get signer() {
    return this.get('signer');
  }
  /**
   * @description The [[Balance]] tip
   */


  get tip() {
    return this.extra.tip;
  }

  injectSignature(signer, signature, _ref) {
    let {
      era,
      nonce,
      tip
    } = _ref;
    this.extra.set('era', era);
    this.extra.set('nonce', nonce);
    this.extra.set('tip', tip);
    this.set('signer', signer);
    this.set('signature', signature);
    return this;
  }
  /**
   * @description Adds a raw signature
   */


  addSignature(signer, signature, payload) {
    return this.injectSignature(new _Address.default(signer), new _Signature.default(signature), new _SignaturePayload.default(payload));
  }
  /**
   * @description Generate a payload and pplies the signature from a keypair
   */


  sign(method, account, _ref2) {
    let {
      blockHash,
      era,
      nonce,
      tip
    } = _ref2;
    const signer = new _Address.default(account.publicKey);
    const payload = new _SignaturePayload.default({
      blockHash,
      era: era || _constants.IMMORTAL_ERA,
      method: method.toU8a(),
      nonce,
      tip: tip || 0
    });
    const signature = new _Signature.default(payload.sign(account));
    return this.injectSignature(signer, signature, payload);
  }
  /**
   * @description Encodes the value as a Uint8Array as per the SCALE specifications
   * @param isBare true when the value has none of the type-specific prefixes (internal)
   */


  toU8a(isBare) {
    return this.isSigned ? super.toU8a(isBare) : new Uint8Array();
  }

}

exports.default = ExtrinsicSignatureV2;
