// @polkadot/types/primitive/Extrinsic/v2/SignaturePayload.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _Struct = _interopRequireDefault(require("../../../codec/Struct"));

var _U8a = _interopRequireDefault(require("../../../codec/U8a"));

var _Hash = _interopRequireDefault(require("../../Hash"));

var _ExtrinsicExtra = require("./ExtrinsicExtra");

var _util = require("../util");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

const basePayload = _objectSpread({}, _ExtrinsicExtra.extraDefinition, {
  blockHash: _Hash.default
});
/**
 * @name SignaturePayloadV2
 * @description
 * A signing payload for an [[Extrinsic]]. For the final encoding, it is variable length based
 * on the contents included
 */


class SignaturePayloadV2 extends _Struct.default {
  constructor(value) {
    super(_objectSpread({
      method: _U8a.default
    }, basePayload), value);
  }
  /**
   * @description The block [[Hash]] the signature applies to (mortal/immortal)
   */


  get blockHash() {
    return this.get('blockHash');
  }
  /**
   * @description The [[U8a]] contained in the payload
   */


  get method() {
    return this.get('method');
  }
  /**
   * @description The [[ExtrinsicEra]]
   */


  get era() {
    return this.get('era');
  }
  /**
   * @description The [[NonceCompact]]
   */


  get nonce() {
    return this.get('nonce');
  }
  /**
   * @description The tip [[BalanceCompact]]
   */


  get tip() {
    return this.get('tip');
  }
  /**
   * @description Sign the payload with the keypair
   */
  // eslint-disable-next-line @typescript-eslint/no-unused-vars


  sign(signerPair) {
    return (0, _util.sign)(signerPair, this.toU8a());
  }

}

exports.default = SignaturePayloadV2;
